includes/class-sensei-enrolment-calculation-scheduler.php

```diff
@@ -0,0 +1,137 @@
+
+	/**
+	 * Number of seconds between each job run.
+	 *
+	 * @var integer
+	 */
+	private $interval;
+
+	/**
+	 * Number of users for each job run.
+	 *
+	 * @var integer
+	 */
+	private $batch_size;
+
+
+	/**
+	 * Sensei_Enrolment_Calculation_Scheduler constructor.
+	 *
+	 * @param integer $interval The scheduler's interval.
+	 * @param integer $batch_size The scheduler's batch size.
+	 */
+	public function __construct( $interval, $batch_size ) {
+		$this->interval   = $interval;
+		$this->batch_size = $batch_size;
+
+		// phpcs:ignore WordPress.WP.CronInterval.ChangeDetected -- Job size small.
+		add_filter( 'cron_schedules', [ $this, 'add_schedule' ] );
+		add_action( 'sensei_calculate_enrolments', [ $this, 'calculate_enrolments' ], 10, 0 );
+		add_action( 'init', [ $this, 'start' ], 101 );
+	}
+
+	/**
+	 * This method starts the scheduler if the run is not completed for the current Sensei version.
+	 */
+	public function start() {
+		if ( get_option( self::CALCULATION_VERSION_OPTION_NAME ) === Sensei_Course_Enrolment_Manager::get_enrolment_calculation_version() ) {
+			return;
+		}
+
+		if ( ! wp_next_scheduled( 'sensei_calculate_enrolments' ) ) {
+			wp_schedule_event( time(), 'sensei_user_recalculation', 'sensei_calculate_enrolments' );
+		}
+	}
+
+	/**
+	 * Stops the scheduler.
+	 */
+	public function stop() {
+		wp_clear_scheduled_hook( 'sensei_calculate_enrolments' );
+	}
+
+
+	/**
+	 * Marks the run as completed and stops the scheduler.
+	 */
+	public function complete() {
+		update_option(
+			self::CALCULATION_VERSION_OPTION_NAME,
+			Sensei_Course_Enrolment_Manager::get_enrolment_calculation_version()
+		);
+		$this->stop();
+	}
+
+	/**
+	 * Adds a custom WP schedule.
+	 *
+	 * @param array $schedules The existing schedules.
+	 *
+	 * @access private
+	 *
+	 * @return array The schedules
+	 */
+	public function add_schedule( $schedules ) {
+		$schedules['sensei_user_recalculation'] = [
+			'interval' => $this->interval,
+			'display'  => esc_html__( 'Every ' . $this->interval . ' seconds.', 'sensei-lms' ), // phpcs:ignore WordPress.WP.I18n.NonSingularStringLiteralText -- This is unavoidable if we want to use a configurable interval.
+		];
+
+		return $schedules;
+	}
+
+	/**
+	 * This method will be called at each scheduled job. It calculates the enrolments for a batch of users.
+	 *
+	 * @access private
+	 */
+	public function calculate_enrolments() {
+
+		$meta_query = [
+			'relation' => 'OR',
+			[
+				'key'     => Sensei_Course_Enrolment_Manager::LEARNER_CALCULATION_META_NAME,
+				'value'   => Sensei_Course_Enrolment_Manager::get_enrolment_calculation_version(),
+				'compare' => '!=',
+			],
+			[
+				'key'     => Sensei_Course_Enrolment_Manager::LEARNER_CALCULATION_META_NAME,
+				'compare' => 'NOT EXISTS',
+			],
+		];
+
+		$user_args = [
+			'fields'     => 'ID',
+			'number'     => $this->batch_size,
+			'meta_query' => $meta_query, // phpcs:ignore  Slow query ok.
```

Comment:
We should ignore the actual rule here.
```suggestion
			'meta_query' => $meta_query, // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query -- Slow query ok.
```

Reply:
Fixed 313886990129312d4e86caa929cc4c87a3c6f6f4